### PR TITLE
basic GroupedInductorBenchmarker

### PR DESCRIPTION
**Pull request by nmacchioni**
**TLDR;** The most basic custom benchmarker is already significantly faster than the default Triton `do_bench` implementation. This is most likely caused by the reduced cache flush size (256MB -> 50MB on A100, which saves ~200MB of bandwidth per flush or ~0.1ms per flush given A100's theoretical peak memory bandwidth ~2TB/s). There is some performance churn, which is probably also caused by the reduced cache flush size, which we will address in later versions.

**Training Results**
Screenshot 2024-12-05 at 12 57 41 PM

**Inference Results**
Screenshot 2024-12-05 at 12 57 54 PM

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #144507
* #144505
* #144501
* #144353
* #133287
* #144365
* __->__ #133121

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov @rec
Screenshot 2024-12-05 at 12 57 54 PM

### CODE REVIEW

**Comment by nmacchioni** [comment]
@eellison do you think we could start landing parts of the stack?

File: torch/_inductor/codegen/multi_kernel.py [hunk 1]
```diff
@@ -367,10 +367,11 @@ def inner():
...
             return inner
 
-        return [
-            benchmarker.benchmark_gpu(wrap_fn(kernel), rep=40)
-            for kernel in self.kernels
-        ]
+        return benchmarker.benchmark_many_gpu(
+            [wrap_fn(kernel) for kernel in self.kernels],
+            rep=40,
+            fast_flush=True,
```
Comment (eellison) [inline torch/_inductor/codegen/multi_kernel.py:373]: Could we remove the `flash_flush` arg here ? theres no reason for it, right ?

File: torch/_inductor/codegen/triton.py [hunk 2]
```diff
@@ -3737,11 +3737,12 @@ def store_cache():
...
                 # generating out of range indices for later calls.
-                ms = benchmarker.benchmark_gpu(
-                    lambda: call(wrapped_jit_function.clone_args(*args)[0])
+                kernel_callable = lambda: call(  # noqa: E731
+                    wrapped_jit_function.clone_args(*args)[0]
                 )
-                ms_clone = benchmarker.benchmark_gpu(
-                    lambda: wrapped_jit_function.clone_args(*args)[0]
+                clone_callable = lambda: wrapped_jit_function.clone_args(*args)[0]  # noqa: E731
+                ms, ms_clone = benchmarker.benchmark_many_gpu(
```
Comment (eellison) [inline torch/_inductor/codegen/triton.py:3744]: benchmark_many_gpu is eventually going to prune between the choices to find the fastest one (right?). in this case, we're not actually comparing these two for perf and dont expect them to give similar results. Should we benchmark separately ? or am i not understanding this correctly

File: torch/_inductor/runtime/benchmarking.py [hunk 3]
```diff
@@ -383,4 +395,123 @@ def benchmark_gpu(
...
+                start_event.record()
+                _callable()
+                end_event.record()
+        torch.cuda.synchronize()
+        estimated_timings = self.get_interleaved_event_pairs_min_timing(
+            interleaved_event_pairs
+        )
+
+        # adjust `benchmark_iters` to fit in the maximum benchmarking duration, we're
+        # alloted `max_benchmark_duration` per-callable, so we can just take the average
```
Comment (eellison) [inline torch/_inductor/runtime/benchmarking.py:478]: nit: /s/alloted/allotted

File: torch/_inductor/runtime/benchmarking.py [hunk 4]
```diff
@@ -383,4 +395,123 @@ def benchmark_gpu(
...
+        # of the estimated timings
+        benchmark_iters = max(
+            min(benchmark_iters, max_benchmark_duration // mean(estimated_timings)), 1
+        )
+
+        # do the memory warmup
+        for _ in range(memory_warmup_iters):
```
Comment (eellison) [inline torch/_inductor/runtime/benchmarking.py:485]: Why do we need to do for this for 100 iterations ? It's not clear to me.

**Review by eellison — COMMENTED** [review]
Looks great ! have a couple questions

File: torch/_inductor/runtime/benchmarking.py (hunk 4 above)
Comment (nmacchioni) [reply torch/_inductor/runtime/benchmarking.py:485]: I picked 100 from testing various iters, it provided the best consistency to overhead ratio

File: torch/_inductor/runtime/benchmarking.py [hunk 5]
```diff
@@ -42,6 +42,25 @@ class Benchmarker:
     def __init__(self: Self) -> None:
         pass
 
+    def infer_device_type(
+        self: Self, fn_args: Tuple[Any, ...], fn_kwargs: Dict[str, Any]
+    ) -> Any:
```
Comment (eellison) [inline torch/_inductor/runtime/benchmarking.py:47]: nit: `-> torch.device`

File: torch/_inductor/runtime/benchmarking.py [hunk 6]
```diff
@@ -42,6 +42,25 @@ class Benchmarker:
     def __init__(self: Self) -> None:
         pass
 
+    def infer_device_type(
+        self: Self, fn_args: Tuple[Any, ...], fn_kwargs: Dict[str, Any]
```
Comment (eellison) [inline torch/_inductor/runtime/benchmarking.py:46]: `fn_kwargs` default of None ?

File: torch/_inductor/runtime/benchmarking.py [hunk 7]
```diff
@@ -130,6 +136,70 @@ def run_for(ms: int) -> List[float]:
...
+
+        Keyword Arguments:
+        - **kwargs: The benchmarking implementation's kwargs.
+
+        Returns:
+        - The runtimes of `fn(*fn_args, **fn_kwargs)`, for `fn`, `fn_args`, and `fn_kwargs` in `fns`,
+        in milliseconds.
+        """
+        inferred_device = None
+        for fn_args, fn_kwargs in zip(fns_args, fns_kwargs):
```
Comment (eellison) [inline torch/_inductor/runtime/benchmarking.py:170]: nit: 

`inferred_device = self.infer_device_type([self.infer_device_type(args, kwargs) for args, kwargs in zip(fns_args, fn_kwargs)], {})` 

maybe

File: torch/_inductor/runtime/benchmarking.py [hunk 8]
```diff
@@ -282,6 +352,129 @@ def benchmark_gpu(
...
+        the values of `memory_warmup_iters` and `benchmark_iters`, along with the
+        estimated runtime of `_callable` and various other factors, and we then
+        shrink `benchmark_iters` to fit in the alloted maximum duration.
+        - **kwargs: Additional kwargs that may be passed to the fallback.
+
+        Returns:
+        - The minimum runtime of each callable in `callables`, in milliseconds.
+        """
+        # we don't want any outside errors propagating into benchmarking
+        torch.cuda.synchronize()
```
Comment (eellison) [inline torch/_inductor/runtime/benchmarking.py:407]: We should use the generic device interface synchronize, as that's what triton do_bench does.

File: torch/_inductor/runtime/benchmarking.py (hunk 3 above)
Comment (eellison) [reply torch/_inductor/runtime/benchmarking.py:434]: ^^

File: torch/_inductor/runtime/benchmarking.py [hunk 9]
```diff
@@ -282,6 +352,129 @@ def benchmark_gpu(
...
+                end_event.record()
+        torch.cuda.synchronize()
+        estimated_timings = self.get_interleaved_event_pairs_min_timing(
+            interleaved_event_pairs
+        )
+
+        # adjust `benchmark_iters` to fit in the maximum benchmarking duration, we're
+        # alloted `max_benchmark_duration` per-callable, so we can just take the average
+        # of the estimated timings
+        benchmark_iters = max(
```
Comment (eellison) [inline torch/_inductor/runtime/benchmarking.py:436]: I guess this runs the risk of very small fn + very large fn running the small fn not very often. should still be fine..

**Review by eellison — COMMENTED** [review]
looks great ! 

I think just need to update to use more generic device interface. Also, cc @blaine-rister if anything here jumps out as being particularly friendly for mtia or should be opted out of.

Also - is it worth just having a basic test that shows square mm of 64 is slower than square mm of 1024 ?
